src/mlpack/methods/ann/layer/policy_impl.hpp

```diff
@@ -0,0 +1,63 @@
+/**
+ * @file policy_impl.hpp
+ * @author Shangtong Zhang
+ *
+ * Implementation of the Policy class.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LAYER_POLICY_IMPL_HPP
+#define MLPACK_METHODS_ANN_LAYER_POLICY_IMPL_HPP
+
+// In case it hasn't yet been included.
+#include "policy.hpp"
+
+namespace mlpack {
+namespace ann {
+
+template<typename InputDataType, typename OutputDataType>
+Policy<InputDataType, OutputDataType>::Policy(
+    double entropyRegularizationWeight) :
+    entropyRegularizationWeight(entropyRegularizationWeight)
+{ /* Nothing to do here. */ }
+
+template<typename InputDataType, typename OutputDataType>
+template<typename InputType, typename OutputType>
+void Policy<InputDataType, OutputDataType>::Forward(
+    const InputType &&input, OutputType &&output)
+{
+  arma::mat maxInput = arma::repmat(arma::max(input), input.n_rows, 1);
+  arma::mat expInput = arma::exp(input - maxInput);
+  arma::mat dense = arma::repmat(arma::sum(expInput), expInput.n_rows, 1);
+  output = expInput / dense;
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void Policy<InputDataType, OutputDataType>::Backward(
+    const arma::Mat<eT> &&prob, arma::Mat<eT> &&advantage, arma::Mat<eT> &&g)
+{
+  arma::mat adv = arma::repmat(arma::sum(advantage), advantage.n_rows, 1);
+  arma::mat gradientPolicy = advantage - adv % prob;
+  arma::mat logProb = arma::log(prob);
```

Comment:
It works well in the unit test. I computed the ground truth gradient by pytorch
```
    import torch
    from torch.autograd import Variable
    import torch.nn.functional as F
    x = Variable(torch.DoubleTensor([[1, 3, 9]]), requires_grad=True)
    prob = F.softmax(x)
    log_prob = F.log_softmax(x)
    policy_loss = -log_prob.gather(1, Variable(torch.LongTensor([[1]]))) * 2
    kl_loss = (prob * log_prob).sum()
    (policy_loss + 0.1 * kl_loss).backward()
    print x.grad.data.numpy()
```
maybe I need to implement same architecture in pytorch to see whether it works or not.